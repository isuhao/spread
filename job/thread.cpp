#include "thread.hpp"

#include <boost/thread.hpp>
#include <assert.h>

using namespace Jobify;

struct ThreadObj
{
  Job *j;
  void operator()()
  {
    j->run();
    assert(j->getInfo()->isFinished());
    delete j;
  }
};

void Thread::sleep(double seconds)
{
  int msecs = (int)(seconds*1000000);
  boost::this_thread::sleep(boost::posix_time::microseconds(msecs));
}

void Thread::run(Job *j, bool async)
{
  ThreadObj to;
  to.j = j;
  if(async) boost::thread trd(to);
  else to();
}
