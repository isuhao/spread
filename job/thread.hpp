#ifndef __JOBIFY_THREAD_HPP
#define __JOBIFY_THREAD_HPP

#include "job.hpp"

namespace Jobify
{
  /* This black-box module swallows any Job object, executes it in a
     separate thread, then deletes the object after it returns. (The
     job must have been allocated using 'new'.)

     You can disable threading (run it locally) by setting
     async=false. This may seem pointless compared to just running
     j->run(), but it does provide a consistent interface for both
     cases, and it will also delete the Job before returning.
   */
  struct Thread
  {
    static void run(Job *j, bool async=true);
    static void sleep(double seconds);
  };
}
#endif
