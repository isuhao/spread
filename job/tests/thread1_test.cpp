#include "thread.hpp"
#include <iostream>
#include <unistd.h>

using namespace Jobify;
using namespace std;

struct MyJob : Job
{
  MyJob(JobInfoPtr j) : Job(j) {}

  void doJob()
  {
    setBusy("Sleeping for 5 seconds");
    info->total = 5;

    for(int i=0; i<5; i++)
      {
        info->current = i;
        Thread::sleep(1);
      }
    info->current = 5;
    setDone();
  }
};

int main()
{
  JobInfoPtr i(new JobInfo), i2(new JobInfo);

  Thread::run(new MyJob(i));
  Thread::run(new MyJob(i2));

  while(!i->isFinished())
    {
      Thread::sleep(0.3);
      cout << "Status: " << i->message << " (" << i->getCurrent() << "/" << i->getTotal() << ")\n";
    }

  return 0;
}
