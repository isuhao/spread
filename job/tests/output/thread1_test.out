Status: Sleeping for 5 seconds (0/5)
Status: Sleeping for 5 seconds (0/5)
Status: Sleeping for 5 seconds (0/5)
Status: Sleeping for 5 seconds (1/5)
Status: Sleeping for 5 seconds (1/5)
Status: Sleeping for 5 seconds (1/5)
Status: Sleeping for 5 seconds (2/5)
Status: Sleeping for 5 seconds (2/5)
Status: Sleeping for 5 seconds (2/5)
Status: Sleeping for 5 seconds (3/5)
Status: Sleeping for 5 seconds (3/5)
Status: Sleeping for 5 seconds (3/5)
Status: Sleeping for 5 seconds (3/5)
Status: Sleeping for 5 seconds (4/5)
Status: Sleeping for 5 seconds (4/5)
Status: Sleeping for 5 seconds (4/5)
Status: Sleeping for 5 seconds (5/5)
